OPL3 Improvements

jpernst opened the pull request:
I've made a few adjustments to the OPL3 music support that I thought would be worth contributing. Feedback welcome of course.

First, a DMX bug from the original game reversed the stereo channels of FM music relative to normal MIDI, so I've enabled the '-reverse' dmx option by default to switch them back. Vanilla behavior can be restored by using a non-default dmx option string.

Second, the OPL stream was being resampled twice, once from the chip native rate to 44.1k, then from 44.1k to whatever the user's actual output rate is. I've cut out the middle step so the OPL music stream runs at the chip native rate and is then resampled by OpenAL to the proper output rate, which should improve quality and reduce needless work.

Finally and perhaps more importantly, I've added support for multiple OPL chips to reduce channel pressure and prevent note-cuts. Other source ports provide this via libADLmidi, which is great, but I didn't want to pull in a whole new dependency right away. Instead the normal DMX code is extended to allow more chip voices. This doesn't make a huge difference in the vanilla game, but many custom maps use much more detailed midis that can easily exhaust the polyphony of a single OPL chip. The max is now 6 chips (54 pseudo-4op voices) which according to the DMXOPL project, should be ample for most any song. If the number of chips is set to 1 (the default) then behavior is still identical to vanilla, so I think this fits in the spirit of limit-removal.

ceski-1 commented on opl/opl.c:
```diff
@@ -32,7 +32,8 @@ static opl_driver_t *drivers[] =
 static opl_driver_t *driver = NULL;
 
-unsigned int opl_sample_rate = 22050;
+static int num_opl_chips = 1;
```
Are multiple `num_opl_chips` variables needed or could a single variable work as well?

jpernst replied:
Shouldn't be too difficult to boil them down to one. I'll just pick one to be the "canonical" one and extern the others.

ceski-1 commented on opl/opl_sdl.c:
```diff
@@ -137,7 +138,23 @@ int OPL_FillBuffer(byte *buffer, int buffer_samples)
+                mix[1] += sample[1];
+            }
+
+            cursor[0] = mix[0] < -32768 ? -32768 : mix[0] > 32767 ? 32767 : mix[0];
+            cursor[1] = mix[1] < -32768 ? -32768 : mix[1] > 32767 ? 32767 : mix[1];
```
For these range checks, consider using the `BETWEEN` macro: `BETWEEN(-32768, 32767, mix[n])`

jpernst replied:
Ah, I was wondering if there was a CLAMP-style macro. I'll add this.

ceski-1 commented on opl/opl_sdl.c:
```diff
@@ -161,8 +178,11 @@ static void OPL_SDL_Shutdown(void)
-static int OPL_SDL_Init(unsigned int port_base)
+static int OPL_SDL_Init(unsigned int port_base, int *num_chips)
 {
+    num_opl_chips = *num_chips < 1 ? 1 : *num_chips > OPL_MAX_CHIPS ? OPL_MAX_CHIPS : *num_chips;
+    *num_chips = num_opl_chips;
```
Same thing here, but maybe changes to the handling of `num_opl_chips` will simplify this anyway.

ceski-1 commented on src/mn_setup.c:
```diff
@@ -2362,6 +2362,9 @@ static setup_menu_t gen_settings2[] = {
      {"midi_player_menu"}, .strings_id = str_midi_player,
      .action = SetMidiPlayer},
 
+    {"Number of OPL Chips", S_THERMO, CNTR_X, M_THRM_SPC, {"num_opl_chips"},
+     .action = SetMidiPlayer},
```
I'd prefer to keep this setting as config-only. If the performance impact isn't too great, a reasonable default that only enthusiasts would ever want to change seems like the way to go. In the future, maybe there will be a separate advanced MIDI menu, and options like this can be reconsidered at that time.

jpernst replied:
I had some difficulty deciphering how the config system works. What exactly is required for a config-only option? Just the Bind_X call?

ceski-1 replied:
A `BIND_NUM` call should work. It's range checked automatically by the min/max values you set. Maybe @rfomin can give some additional tips.

rfomin replied:
Yes, just `BIN_NUM` macro will do it.

> there will be a separate advanced MIDI menu

I like the idea!

ceski-1 reviewed (commented):
It's always nice to see music improvements. I'm not as familiar with the OPL3 code, so I'll share some high-level feedback. Maybe others would like to comment as well.

> The max is now 6 chips (54 pseudo-4op voices) which according to the DMXOPL project, should be ample for most any song.

Microsoft GS Wavetable Synth, a typical baseline MIDI target, only has a polyphony of 32, so 6 chips max seems reasonable to me.

I'm curious about the performance impact of simulating multiple chips. If it's not too great, maybe it makes sense to set a higher default (e.g. 4 chips for 36 voices).

rfomin reviewed (commented):
Thank you, looks very good! Could you recommend MIDIs for testing?

rfomin commented:
> I'm curious about the performance impact of simulating multiple chips.

It's quite heavy on my old laptop. On midern CPU it's fine.

jpernst commented:
> Thank you, looks very good! Could you recommend MIDIs for testing?

I noticed the problem particularly in Thatcher's Techbase map02. On a single chip it drops notes like crazy.

As for CPU load, the Nuked OPL core used here is known to be very accurate, but CPU heavy. There other other cores that are almost as accurate but much faster that might be worth exploring if proves to be a problem.

fabiangreffrath commented on src/i_oplmusic.c:
```diff
@@ -359,7 +359,7 @@ static unsigned int last_perc_count;
 // adlib chip.
 
-static char *snd_dmxoption = "-opl3"; // [crispy] default to OPL3 emulation
+static char *snd_dmxoption = "-opl3-reverse"; // [crispy] default to OPL3 emulation
```
Please revert this. I've had this in Crispy for a while and people reported issues when the music didn't sound as they were used to anymore. 

Sometimes "technically accurate" and "Vanilla" have contrary means, and in these cases we tend to decide for "Vanilla".

jpernst commented:
OK, I took another pass and addressed most of the feedback I think. It feels pretty clean now, just needs a bit of stress testing on any elaborate MIDIs anyone can think of. Speaking of which, if you add DMXOPL.wad to your all-all autoload, you'll get a better set of patches that will also help stress test the mult-chip system since they use more voices than the vanilla patches do.

rfomin reviewed (approved):
Looks good to me, thank you!

> if you add DMXOPL.wad to your all-all autoload, you'll get a better set of patches that will also help stress test the mult-chip system since they use more voices than the vanilla patches do.

That makes a big difference! It's Creative Commons, I think we can add it to the Woof distribution with an option in the future "MIDI Options" secondary menu.

ceski-1 reviewed (approved):
Thank you!

ceski-1 commented:
> DMXOPL.wad

Maybe we should add a link to it in the next changelog.

fabiangreffrath commented:
Is it possible to dynamically allocate more chips depending on the number of voices the current song requires?

jpernst commented:
Hmm, actually yes. Thinking about it that should be pretty simple. I can simply leave the higher chips in a suspended state until they receive a chip event. Should be easy to do; I'll look into adding that tomorrow

rfomin commented:
> Is it possible to dynamically allocate more chips depending on the number of voices the current song requires?

Nuked OPL is pretty heavy. On my 10+ year old laptop it's starting to skip the buffer with 6 chips. Maybe if it's only optional.

jpernst commented:
I plan to do it so that the config variable specifies that maximum number of chips that are allowed to run, just as it does now, but chips that haven't yet been actually used will remain inactive. This way you can set a CPU-ceiling that your system can tolerate, but simpler MIDIs will benefit from less overhead.
